### ISSUE TITLE
Use another backends for check, type info, autocompletion, etc.

### ISSUE THREAD

**Issue by chaika2013** [atom-haskell/ide-haskell#14]

**Comment by ijuma** [comment]
https://www.fpcomplete.com/blog/2015/03/announce-ide-backend is interesting and there is mention of Atom support there (I think https://github.com/edsko/atom-cabal uses it).

**Comment by lierdakil** [comment]
It sounds interesting, but I don't think it's ready for public consumption
yet. Worth investigating though, so thank you for heads-up.

2015-03-30 11:49 GMT+03:00 Ismael Juma <notifications@github.com>:

**Comment by ijuma** [comment]
Agreed that it's still early days. Worth keeping an eye on the JSON client, in particular:

https://github.com/chrisdone/ide-backend-client

**Comment by JorisM** [comment]
has anyone looked into using ide-backend yet? I might give it a try...

**Comment by ijuma** [comment]
@JorisM, I don't think anyone has done it, so go for it. :)

**Comment by lierdakil** [comment]
Please hold off on that for a bit. I'm currently working on moving ghc-mod support to separate package, which would allow for easier maintenance of backend providers. But since API is kind of a jumbled mess at the moment, that's not something I am comfortable sharing yet.

In a bit, when I've got backend service API stable and published, it'll be easier to implement ide-backend as a backend for ide-haskell.

If you're eager to start hacking away, see [haskell-ghc-mod's services][1] service providers for a general idea of what API might look like. It is subject to sudden change though.

[1]: https://github.com/atom-haskell/haskell-ghc-mod/blob/master/lib/haskell-ghc-mod.coffee#L104

**Comment by lierdakil** [comment]
Ok, a little update. While I feel quite comfortable with current implementation of `haskell-ide-backend` service, `haskell-completion-backend` service needs some serious work. I think that it might benefit from public discussion. Any suggestions are welcome.

Meanwhile, I looked over how `ide-backend` handles autocompletion, and I have to say I'm not very impressed, since it seems to completely lack any context sensitivity, offering completions based on just file and prefix. That seems insufficient, at least for what I would ultimately want with autocompletion. A couple examples, which would depend on context sensitivity:
1. Pragmas (language extensions, GHC options)
2. import-clause completions (module name, hiding/import list)
3. Typeclass completions (deriving, instance, polymorphic functions, ...)
4. etc

**Comment by archaeron** [comment]
So your plan is to make all the different parts of the backends work as providers?
This is really great news.

**Comment by lierdakil** [comment]
@archaeron, well, the general plan is to have `ide-haskell` as a graphical frontend, and `haskell-ghc-mod` (and possibly other packages in the future) as a backend, communicating via service-hub. This would allow to mix and match backend providers and frontend interfaces. That said, I will likely primarily focus on `ide-haskell` and `haskell-ghc-mod`, so other backends/frontends would have to be a bit of a community effort.

**Comment by archaeron** [comment]
sounds like a great idea! I'm looking forward to the next version

**Comment by lierdakil** [comment]
Okay, I have released ide-haskell-0.5.0. and haskell-ghc-mod-0.6.0. Latter now has annotated reference implementations of backend services (see [here][1]). More detailed documentation will follow. If you are interested in writing a new backend implementation, feel free to contact me via e-mail/gitter/new issue for questions/endorsement/etc.

I will close this issue as out-of-scope now.

[1]: https://github.com/atom-haskell/haskell-ghc-mod

**Comment by cies** [comment]
> Meanwhile, I looked over how ide-backend handles autocompletion, and I have to say I'm not very impressed, since it seems to completely lack any context sensitivity, offering completions based on just file and prefix. That seems insufficient, at least for what I would ultimately want with autocompletion. A couple examples, which would depend on context sensitivity:
> 1. Pragmas (language extensions, GHC options)
> 2. import-clause completions (module name, hiding/import list)
> 3. Typeclass completions (deriving, instance, polymorphic functions, ...)
> 4. etc

@lierdakil: Did you contact @chrisdone about the features you miss in `ide-backend`? Maybe he has them planned...

Apart from these missing features I think that `ide-backend` is a rather good match with `ide-haskell` as JSON is used as protocol.

**Comment by lierdakil** [comment]
@cies, `ide-backend` itself uses something weird for protocol, I didn't dig deeper. `ide-backend-client` translates that to json.

No, I haven't contacted anyone on this topic, because I'm not currently interested in using `ide-backend`, and I have my hands full as it is.

**Comment by chrisdone** [comment]
You're too busy reproducing everything in ide-backend to use ide-backend. ;-)

**Comment by lierdakil** [comment]
@chrisdone, I really don't see what `ide-backend` offers besides autocompletion (with which I have gripes, as outlined above).

**Comment by cies** [comment]
@lierdakil: my wrong, I meant `ide-backend-client`.

> You're too busy reproducing everything in ide-backend to use ide-backend. ;-)

@chrisdone: I'm not sure I understand your remark. Just checked [your personal emacs config](https://github.com/chrisdone/chrisdone-emacs) and was delighted to see it uses ide-backend-mode since one month now.

**Comment by cies** [comment]
@lierdakil, I [found the following requests types](https://github.com/chrisdone/ide-backend-client/blob/master/src/IdeSession/Client/JsonAPI.hs) in `ide-backend-client`:

    RequestGetSourceErrors
    RequestGetLoadedModules
    RequestGetSpanInfo SourceSpan
    RequestGetExpTypes SourceSpan
    RequestGetAutocompletion AutocompletionSpan

From there I see it has some to offer besides autocomplete. What functionality would be required for you to consider it's feature set complete?

**Comment by lierdakil** [comment]
@cies, let me rephrase that. I don't see what `ide-backend` offers besides autocompletion *compared to what I already use* (i.e. `ghc-mod`). Sorry about poor wording, I was in a bit of hurry at the time.

**Comment by cies** [comment]
@lierdakil, thanks for clarifying.
